src/sentry_sync.c

```diff
@@ -41,78 +35,95 @@ typedef struct sentry_bgworker_task_s {
     void *task_data;
 } sentry_bgworker_task_t;
 
+static void
+sentry__task_incref(sentry_bgworker_task_t *task)
+{
+    sentry__atomic_fetch_and_add(&task->refcount, 1);
+}
+
+static void
+sentry__task_decref(sentry_bgworker_task_t *task)
+{
+    if (sentry__atomic_fetch_and_add(&task->refcount, -1) == 1) {
+        if (task->cleanup_func) {
+            task->cleanup_func(task->task_data);
+        }
+        sentry_free(task);
+    }
+}
+
 struct sentry_bgworker_s {
-    sentry_cond_t submit_signal;
-    sentry_mutex_t submit_signal_lock;
-    sentry_cond_t done_signal;
-    sentry_mutex_t done_signal_lock;
-    sentry_mutex_t task_lock;
     sentry_threadid_t thread_id;
+    sentry_cond_t task_signal;
+    sentry_mutex_t task_lock;
     sentry_bgworker_task_t *first_task;
     sentry_bgworker_task_t *last_task;
-    size_t task_count;
     void *state;
     void (*free_state)(void *state);
     long refcount;
-    bool running;
+    long running;
```

Comment:
I did some more digging, and apparently it is discouraged. In this case, I'm thinking that `volatile bool` should be sufficient to toggle a flag in another thread *eventually* (that is, without atomicity constraints), but since it is discouraged, let's maybe stick with your atomic long. WDYT?

Reply:
The comments on [this](https://stackoverflow.com/a/4558024/4228225) stackoverflow answer pretty much capture it. Funny enough, the figure that our case is the one where `volatile` would make sense, but `atomic` is the better choice. Given that we have atomic in now, let's keep it, even though it looks more complicated.